### ISSUE TITLE
Threads Spin Indefinitely at java.util.WeakHashMap.get 

### ISSUE THREAD

**Issue by cjet** [oracle/graal#1916]
GraalVM Version: 19.2.1
Java Version: 11.0.4

My application uses GraalVM to run JavaScript code snippets. I have 8 threads running JavaScript code. Each of these 8 threads has its own context, but shares an engine. In the thread dump, these threads have a name that starts with "JSRunner-".

There are also another 8 watchdog threads. These threads monitor the execution of the JavaScript code and forcefully close the context if the execution exceeds a time limit. These threads have a name that start with "watchdog-thread-". 

On several occasions, I have seen the application become unresponsive because each of the "JSRunner-1" threads are stuck with a stack similar to the following:

> "JSRunner-1" #413 daemon prio=5 os_prio=0 cpu=1218834.75ms elapsed=2861.83s tid=0x00007fb54801b000 nid=0x9742 runnable  [0x00007fb50b6f2000]
   java.lang.Thread.State: RUNNABLE
	at java.util.WeakHashMap.get(java.base@11.0.4/WeakHashMap.java:404)
	at java.util.Map.computeIfAbsent(java.base@11.0.4/Map.java:1001)
	at com.oracle.truffle.js.runtime.objects.JSShapeData.getShapeData(JSShapeData.java:109)
	at com.oracle.truffle.js.runtime.objects.JSShapeData.getEnumerablePropertyNamesArray(JSShapeData.java:136)
	at com.oracle.truffle.js.runtime.objects.JSShapeData.getEnumerablePropertyNames(JSShapeData.java:146)
	at com.oracle.truffle.js.runtime.objects.JSShape.getEnumerablePropertyNames(JSShape.java:196)

Additional, all of the watchdog threads have a stack similar to the following:

> "watchdog-thread-1" #425 prio=5 os_prio=0 cpu=84.25ms elapsed=2859.07s tid=0x00007fb82c048000 nid=0x977b in Object.wait()  [0x00007fb50a4f1000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(java.base@11.0.4/Native Method)
	- waiting on <no object reference available>
	at com.oracle.truffle.polyglot.PolyglotContextImpl.waitForClose(org.graalvm.truffle/PolyglotContextImpl.java:905)
	- waiting to re-lock in wait() <0x0000000571b142a0> (a com.oracle.truffle.polyglot.PolyglotContextImpl)
	at com.oracle.truffle.polyglot.PolyglotEngineImpl$CancelHandler.waitForClosing(org.graalvm.truffle/PolyglotEngineImpl.java:1055)
	at com.oracle.truffle.polyglot.PolyglotContextImpl.close(org.graalvm.truffle/PolyglotContextImpl.java:858)
	at org.graalvm.polyglot.Context.close(org.graalvm.sdk/Context.java:635)

It appears the "JSRunner-" threads are in an infinite loop at `java.util.WeakHashMap.get`. If access to the map is not appropriately protected in a multithreaded context, race conditions and data corruption can occur. A similar issue is detailed here: https://mailinator.blogspot.com/2009/06/beautiful-race-condition.html 

A partial thread dump from my application is attached. 

[partial_thread_dump.log](https://github.com/oracle/graal/files/3913562/partial_thread_dump.log)

**Comment by dsmolyarenko** [comment]
GraalVM Version: 19.3.0
Java Version: 11.0.2

Similar problem. Single engine is being shared between twelve thread local polyglot contexts which are always kept opened for better performance.

typical thread traces when stuck:
```
	java.base@11.0.2/java.util.WeakHashMap.get(WeakHashMap.java:404)
	java.base@11.0.2/java.util.Map.computeIfAbsent(Map.java:1001)
	app//com.oracle.truffle.js.runtime.objects.JSShapeData.getShapeData(JSShapeData.java:109)
	app//com.oracle.truffle.js.runtime.objects.JSShapeData.getEnumerablePropertyNamesArray(JSShapeData.java:136)
	app//com.oracle.truffle.js.runtime.objects.JSShapeData.getEnumerablePropertyNames(JSShapeData.java:146)
	app//com.oracle.truffle.js.runtime.objects.JSShape.getEnumerablePropertyNames(JSShape.java:196)
	app//com.oracle.truffle.js.runtime.objects.JSObject.enumerableOwnNames(JSObject.java:473)
	app//com.oracle.truffle.js.runtime.builtins.JSClass$GetMembers.nonArrayCached(JSClass.java:478)
	app//com.oracle.truffle.js.runtime.builtins.JSClassGen$InteropLibraryExports$Cached.getMembers(JSClassGen.java:149)
	app//com.oracle.truffle.api.interop.InteropLibrary.getMembers(InteropLibrary.java:549)
	app//com.oracle.truffle.polyglot.PolyglotMap$Cache$EntrySet.doCached(PolyglotMap.java:515)
	app//com.oracle.truffle.polyglot.PolyglotMapFactory$CacheFactory$EntrySetNodeGen.executeImpl(PolyglotMapFactory.java:198)
	app//com.oracle.truffle.polyglot.HostToGuestRootNode.execute(HostToGuestRootNode.java:94)
	app//com.oracle.truffle.api.impl.DefaultCallTarget.call(DefaultCallTarget.java:102)
	app//com.oracle.truffle.polyglot.PolyglotMap.entrySet(PolyglotMap.java:117)
	java.base@11.0.2/java.util.AbstractMap.size(AbstractMap.java:85)
	java.base@11.0.2/java.util.AbstractMap.isEmpty(AbstractMap.java:95)
```

```
	java.base@11.0.2/java.util.WeakHashMap.get(WeakHashMap.java:404)
	java.base@11.0.2/java.util.Map.computeIfAbsent(Map.java:1001)
	app//com.oracle.truffle.js.runtime.objects.JSShapeData.getShapeData(JSShapeData.java:109)
	app//com.oracle.truffle.js.runtime.objects.JSShapeData.getEnumerablePropertyNamesArray(JSShapeData.java:136)
	app//com.oracle.truffle.js.runtime.objects.JSShapeData.getEnumerablePropertyNames(JSShapeData.java:146)
	app//com.oracle.truffle.js.runtime.objects.JSShape.getEnumerablePropertyNames(JSShape.java:196)
	app//com.oracle.truffle.js.nodes.access.EnumerableOwnPropertyNamesNode.enumerableOwnPropertyNames(EnumerableOwnPropertyNamesNode.java:96)
	app//com.oracle.truffle.js.nodes.access.EnumerableOwnPropertyNamesNodeGen.execute(EnumerableOwnPropertyNamesNodeGen.java:22)
	app//com.oracle.truffle.js.builtins.ObjectFunctionBuiltins$ObjectKeysNode.enumerableOwnPropertyNames(ObjectFunctionBuiltins.java:799)
	app//com.oracle.truffle.js.builtins.ObjectFunctionBuiltins$ObjectKeysNode.keysDynamicObject(ObjectFunctionBuiltins.java:711)
	app//com.oracle.truffle.js.builtins.ObjectFunctionBuiltinsFactory$ObjectKeysNodeGen.execute(ObjectFunctionBuiltinsFactory.java:1595)
	app//com.oracle.truffle.js.nodes.function.FunctionRootNode.executeInRealm(FunctionRootNode.java:147)
	app//com.oracle.truffle.js.runtime.JavaScriptRealmBoundaryRootNode.execute(JavaScriptRealmBoundaryRootNode.java:93)
	app//com.oracle.truffle.api.impl.DefaultCallTarget.callDirectOrIndirect(DefaultCallTarget.java:85)
	app//com.oracle.truffle.api.impl.DefaultDirectCallNode.call(DefaultDirectCallNode.java:59)
	app//com.oracle.truffle.js.nodes.function.JSFunctionCallNode$UnboundJSFunctionCacheNode.executeCall(JSFunctionCallNode.java:1250)
	app//com.oracle.truffle.js.nodes.function.JSFunctionCallNode.executeCall(JSFunctionCallNode.java:233)
```
Thread CPU tick counter gains as it's 100% loaded. Appears on application start every twentieth run.

**Comment by pmezrin13** [comment]
@woess Could you clarify, please - Is that issue considered as a bug or misusing of Graal API?

**Comment by dsmolyarenko** [comment]
> GraalVM Version: 19.3.0
> Java Version: 11.0.2
> 
> Similar problem. Single engine is being shared between twelve thread local polyglot contexts which are always kept opened for better performance.

According to source code the JSContext (currently one per Engine) that's designed to be used in multi-thread mode is not fully thread safe. Especially the shape data by shape map is accessible from different threads by write operation:
```
public class JSContext {
...
    private synchronized Map<Shape, JSShapeData> createShapeDataMap() {
        Map<Shape, JSShapeData> map = shapeDataMap;
        if (map == null) {
            map = new WeakHashMap<>(); // Collections.synchronizedMap(new WeakHashMap<>());
            shapeDataMap = map;
        }
        return map;
    }
...
```
Concurrent modification is 100% reproducible. Race condition happens mostly on map expanding only. The additional **synchronization** eliminates the problem:
```
            map = Collections.synchronizedMap(new WeakHashMap<>());
```

The **Engine** per thread also **fixes** the issue () but in such case the [code-caching-across-multiple-contexts](https://www.graalvm.org/docs/reference-manual/embed/#code-caching-across-multiple-contexts) won't work.

**Comment by woess** [comment]
Thanks for the report. You're correct, it's a bug that occurs when you share an Engine between multiple threads. The bug is fixed in GraalVM version 19.3.1.
